utils_cv/detection/model.py

```diff
@@ -285,63 +334,122 @@ def plot_precision_loss_curves(
+        **kwargs,
+    ) -> Dict:
+        """ Return prediction results that are above the threshold if any. """
+        # detach prediction results to cpu
+        pred = {k: v.detach().cpu().numpy() for k, v in pred.items()}
+        # apply score threshold
+        if threshold:
+            selected = pred['scores'] > threshold
+            pred = {k: v[selected] for k, v in pred.items()}
+        if "masks" in pred and mask_threshold:
+            pred["masks"] = pred["masks"] > mask_threshold
+        return pred
+
+    def _get_det_bboxes_and_mask(
+        self,
+        pred: Dict,
+        im_path: Union[str, Path]
+    ) -> Dict:
+        """ Extract bounding boxes and masks from the prediction object. """
+
+        label_names = np.array(self.labels)[pred['labels'] - 1]
+        res = {
+            "det_bboxes":
+                DetectionBbox.from_arrays(
+                    pred['boxes'].tolist(),
+                    score=pred['scores'].tolist(),
+                    label_idx=pred['labels'].tolist(),
+                    label_name=label_names.tolist(),
+                    im_path=im_path,
+                )
+        }
+
+        if "masks" in pred:
+            res["masks"] = pred["masks"].squeeze(1)
+        return res
+
     def predict(
         self,
         im_or_path: Union[np.ndarray, Union[str, Path]],
         threshold: Optional[int] = 0.5,
-    ) -> List[DetectionBbox]:
+        **kwargs,
+    ) -> Dict:
         """ Performs inferencing on an image path or image.
 
         Args:
-            im_or_path: the image array which you can get from `Image.open(path)` OR a
-            image path
+            im_or_path: the image array which you can get from
+                `Image.open(path)` OR a image path
             threshold: the threshold to use to calculate whether the object was
             detected. Note: can be set to None to return all detection bounding
             boxes.
 
-        Raises:
-            TypeError is the im object is a path or str to the image instead of
-            an nd.array
-
         Return a list of DetectionBbox
         """
-        im = (
-            Image.open(im_or_path)
+        im, im_path = (
+            (Image.open(im_or_path), im_or_path)
             if isinstance(im_or_path, (str, Path))
-            else im_or_path
+            else (im_or_path, None)
         )
-
-        transform = transforms.Compose([transforms.ToTensor()])
-        im = transform(im).cuda()
+        im = self._transform(im)
         model = self.model.eval()  # eval mode
         with torch.no_grad():
             pred = model([im])
 
-        labels = self.dataset.labels if self.dataset else self.labels
-        det_bboxes = _get_det_bboxes(pred, labels=labels)
-
-        # limit to threshold if threshold is set
-        return _apply_threshold(det_bboxes, threshold)
+        pred = [
+            self._apply_threshold(p, threshold, **kwargs) for p in pred
```

Comment:
Not sure why pred is a list. I.e. we first build a list here, but then in the next line only use pred[0]?

Reply:
Given that we only have a single image, why would there be more than 1 item in pred?

Reply:
The pre-trained PyTorch mask rcnn model needs inputs as list even if there is only one image.  I have made it simpler.

Reply:
I see. Maybe just do pred = model([im])[0] then. In one line we create a list of im, and then take the 0th element.

Reply:
Changed.